File: CompanionApps/iOS/Sources/Handlers/GestureHandler.swift
/// Handles gesture-related gRPC requests by delegating to XCUITestBridge.
final class GestureHandler: @unchecked Sendable {
    private let bridge: XCUITestBridge

    init(bridge: XCUITestBridge) {
        self.bridge = bridge
    }

    func swipe(fromX: Double, fromY: Double, toX: Double, toY: Double, durationMs: Int) async {
        await bridge.swipe(
            fromX: fromX, fromY: fromY,
            toX: toX, toY: toY,
            durationSeconds: Double(durationMs) / 1000.0
        )
    }

    func drag(
        fromX: Double,
        fromY: Double,
        toX: Double,
        toY: Double,
        durationMs: Int,
        holdMs: Int
    ) async {
        await bridge.drag(
            fromX: fromX, fromY: fromY,
            toX: toX, toY: toY,
            durationSeconds: Double(durationMs) / 1000.0,
            holdSeconds: Double(holdMs) / 1000.0
        )
    }

    func scroll(direction: ScrollDirection, distance: Double) async {
        await bridge.scroll(direction: direction, distance: distance)
    }

    /// Returns `false` when `id`/`label`/`containsText` was supplied but resolved no element, so
    /// the caller can surface a real failure instead of reporting success for a swipe that either
    /// did nothing or — previously — silently landed on the wrong element. See
    /// `XCUITestBridge.swipeInDirection` for why a resolution miss must not fall back to a
    /// generic whole-target swipe.
    @discardableResult
    func swipeInDirection(
        direction: ScrollDirection,
        id: String?,
        label: String?,
        containsText: String?
    ) async -> Bool {
        await bridge.swipeInDirection(direction, id: id, label: label, containsText: containsText)
    }
}

File: CompanionApps/Android/app/src/androidTest/java/com/manman/companion/handlers/GestureHandler.kt
package com.manman.companion.handlers

import com.manman.companion.bridge.Direction
import com.manman.companion.bridge.UIAutomatorBridge

class GestureHandler(private val bridge: UIAutomatorBridge) {

    fun swipe(fromX: Int, fromY: Int, toX: Int, toY: Int, durationMs: Int): Boolean {
        val steps = (durationMs / 5).coerceAtLeast(1)
        return bridge.swipe(fromX, fromY, toX, toY, steps)
    }

    fun drag(fromX: Int, fromY: Int, toX: Int, toY: Int, durationMs: Int, holdMs: Int): Boolean {
        return bridge.drag(fromX, fromY, toX, toY, durationMs, holdMs)
    }

    fun scroll(direction: Direction, distance: Int): Boolean {
        return bridge.scroll(direction, distance)
    }

    fun swipeInDirection(
        direction: Direction,
        distance: Int,
        durationMs: Int,
        resourceId: String?,
        text: String?
    ): Boolean {
        return bridge.swipeInDirection(direction, distance, durationMs, resourceId, text)
    }
}
